### PR TITLE
Add `Unsupported` to `std::io::ErrorKind`

### PR DESCRIPTION
**Pull request by CDirkx**
I noticed a significant portion of the uses of `ErrorKind::Other` in std is for unsupported operations.
The notion that a specific operation is not available on a target (and will thus never succeed) seems semantically distinct enough from just "an unspecified error occurred", which is why I am proposing to add the variant `Unsupported` to `std::io::ErrorKind`. 

**Implementation**:

The following variant will be added to `std::io::ErrorKind`:
 
```rust
/// This operation is unsupported on this platform.
Unsupported
```
`std::io::ErrorKind::Unsupported` is an error returned when a given operation is not supported on a platform, and will thus never succeed; there is no way for the software to recover. It will be used instead of `Other` where appropriate, e.g. on wasm for file and network operations. 

`decode_error_kind` will be updated  to decode operating system errors to `Unsupported`:
- Unix and VxWorks: `libc::ENOSYS`
- Windows: `c::ERROR_CALL_NOT_IMPLEMENTED`
- WASI: `wasi::ERRNO_NOSYS`

**Stability**:
This changes the kind of error returned by some functions on some platforms, which I think is not covered by the stability guarantees of the std? User code could depend on this behavior, expecting `ErrorKind::Other`, however the docs already mention:

> Errors that are `Other` now may move to a different or a new `ErrorKind` variant in the future. It is not recommended to match an error against `Other` and to expect any additional characteristics, e.g., a specific `Error::raw_os_error` return value.

The most recent variant added to `ErrorKind` was `UnexpectedEof` in `1.6.0` (almost 5 years ago), but `ErrorKind` is marked as `#[non_exhaustive]` and the docs warn about exhaustively matching on it, so adding a new variant per se should not be a breaking change.

The variant `Unsupported` itself could be marked as `#[unstable]`, however, because this PR also immediately uses this new variant and changes the errors returned by functions I'm inclined to agree with the others in this thread that the variant should be insta-stabilized.

### CODE REVIEW

File: library/std/src/sys/unsupported/os.rs [hunk 1]
```diff
@@ -80,11 +80,11 @@ pub fn getenv(_: &OsStr) -> io::Result<Option<OsString>> {
 }
 
 pub fn setenv(_: &OsStr, _: &OsStr) -> io::Result<()> {
-    Err(io::Error::new(io::ErrorKind::Other, "cannot set env vars on this platform"))
+    Err(io::Error::new(io::ErrorKind::NotSupported, "cannot set env vars on this platform"))
 }
 
 pub fn unsetenv(_: &OsStr) -> io::Result<()> {
-    Err(io::Error::new(io::ErrorKind::Other, "cannot unset env vars on this platform"))
+    Err(io::Error::new(io::ErrorKind::NotSupoorted, "cannot unset env vars on this platform"))
```
Comment (jyn514) [inline library/std/src/sys/unsupported/os.rs:87]:
```suggestion
    Err(io::Error::new(io::ErrorKind::NotSupported, "cannot unset env vars on this platform"))
```

**Comment by scottmcm** [comment]
cc @rust-lang/libs explicitly for this -- this might need to go in insta-stable, since while the variant can be marked unstable, the change to stop emitting the previous one would not be feature-gated.

**Comment by the8472** [comment]
Should `decode_error_kind` be extended to translate any of `libc::{ENOSYS,ENOTSUP,EOPNOTSUPP}` to this ErrorKind?

**Comment by CDirkx** [comment]
I updated `decode_error_kind` to support `NotSupported` on various platforms:

- Unix and VxWorks: `libc::{ENOSYS, ENOTSUP, EOPNOTSUPP}`
- Windows: `c::{ERROR_CALL_NOT_IMPLEMENTED, E_NOTIMPL}`
- CloudABI: `abi::errno::{NOSYS, NOTSUP}` (has no `OPNOTSUPP`)
- WASI: `wasi::{ERRNO_NOSYS, ERRNO_NOSUP}` (has no `OPNOTSUPP`)
- Hermit: (has hard coded error numbers, I could not find a relevant error number)
- SGX: (has no equivalent error in `fortanix_sgx_abi::Error`)

I didn't include errors like `EAFNOSUPPORT` (address family not supported) and `EPROTONOSUPPORT` (protocol not supported) for now.

**Comment by m-ou-se** [comment]
> user code could depend on this behavior, expecting `ErrorKind::Other`

The documentation fortunately already notes that these kind of changes might happen in the future:

> Errors that are `Other` now may move to a different or a new `ErrorKind` variant in the future. It is not recommended to match an error against `Other` and to expect any additional characteristics, e.g., a specific `Error::raw_os_error` return value.

(In hindsight it might have been better if matching 'other' errors against `ErrorKind::Other` was not possible. E.g. by using an unstable/doc(hidden) variant for those. But I guess it's too late for that now.)

I'm not sure if adding  an `#[unstable]` variant to this stable enum is a good idea. Especially as these errors used a `#[stable]` variant before, like @scottmcm mentioned.

> CloudABI: `abi::errno::{NOSYS, NOTSUP}` (has no `OPNOTSUPP`)

Feel free to ignore CloudABI, as that platform itself will be NOTSUP soon: https://github.com/rust-lang/rust/pull/78439

**Comment by KodrAus** [comment]
I can see the value in wanting to check whether or not an IO operation is supported so you can fallback to alternatives.

I also think we should consider marking this as insta-stable if we want to add it. In either case I think an FCP would probably be a good idea before we land it.

**Comment by CDirkx** [comment]
I have updated the PR to immediately stabilize `NotSupported` and updated the description to include the information inthis thread.

**Comment by KodrAus** [comment]
@rfcbot fcp merge

This proposes adding a `NotSupported` variant to `io::ErrorKind` that's used when an operation isn't supported. Callers may want to try falling back to alternatives or bubble up.

File: library/std/src/io/error.rs [hunk 2]
```diff
@@ -178,6 +178,10 @@ pub enum ErrorKind {
     /// read.
     #[stable(feature = "read_exact", since = "1.6.0")]
     UnexpectedEof,
+
+    /// This operation is not supported on this platform.
+    #[stable(feature = "not_supported_error", since = "1.50.0")]
```
Comment (KodrAus) [inline library/std/src/io/error.rs:183]: Just a note that we'll need to bump this `since` field if the FCP completes

**Comment by BurntSushi** [comment]
Could we do a crater run with this change? I know this change is technically okay, but I think it would still be a good idea to get an understanding of what kind of breakage this might cause, if any.

**Comment by jyn514** [comment]
@bors try

(`try` doesn't take much time even if we decide not to do a crater run)

**Comment by Mark-Simulacrum** [comment]
> Could we do a crater run with this change? I know this change is technically okay, but I think it would still be a good idea to get an understanding of what kind of breakage this might cause, if any.

We can, but it seems unlikely to turn anything up. Crater runs with modern linux kernels (5.3 and 5.4 on the two builders I checked), and so it seems unlikely that the test results will have any major cases of unsupported operations.

**Comment by BurntSushi** [comment]
@Mark-Simulacrum Yeah, I was thinking of case analysis failures. But it looks like this enum is tagged with `#[non_exhaustive]`, so yeah, probably won't turn anything up.

**Comment by ijackson** [comment]
I approve of adding more ErrorKinds, so sorry to throw a spanner in the works here.  But I don't think the set of errno values on Unix is right.

Also in general I question whether it is a good idea to bundle different errno values together into one ErrorKind.  Experienced Unix system administrators recognise errno values and the corresponding strings and can often infer useful information from quite subtle distinctions.

In detail:

 * `ENOSYS` means specifically that the system call being called is not implemented.  In practice this probably means that the executable is running on too-old a kernel, or that there is some kind of libc incompatibility.  `ENOSYS` is not typically used for situations when a specific system call parameter value is not supported.
 * `EOPNOTSUPP` means specifically that a socket operation was not supporrted.  This much more specific error generally means that the sequence of system calls on a socket was not appropriate; for example, calling `accept` on a connected socket.  But it can mean other specific things depending on what precisely the fd refers to.

Note that there are a lot of other errno values that refer to more specific kinds of unsupportedness.

I am OK with merging `EOPNOTSUP` and `ENOTSUP` (SuS says they may be the same).

I think for this MR the best answer is probably to include only `ENOSYS` and leave other errno values for future extensions to `ErrorKind`.  This narrowing may warrant a slight change to the description.

**Comment by the8472** [comment]
> Experienced Unix system administrators recognise errno values and the corresponding strings and can often infer useful information from quite subtle distinctions.

`ErrorKind` represents a cross-platform rust side representation of an error that is useful to `match` on. The errno is not discarded, you can query and print it directly if you want more fine-grained, platform-specific information.

**Comment by ijackson** [comment]
> `ErrorKind` represents a cross-platform rust side representation of an error that is useful to `match` on. The errno is not discarded, you can query and print it directly if you want more fine-grained, platform-specific information.

Yes.  But smooshing together errors which can be usefully distinguished, can make it harder to write portable code which knows a lot about the details of a particular OS API area.  This is particularly true of stuff relating to sockets and networking, where the influence of the BSD socket API and therefore of Unix errno values is strong.

I would prefer to avoid asking people to write code which, for each platform, imports `os::...`, uses `into_raw`, and matches on whatever that platform's spelling is for the socket error `EOPNOTSUPP`.  If we provide a distinct `ErrorKind` for that, we would save them that effort.  And we have a good place to put that code.

**Comment by ijackson** [comment]
I guess at this point I should mention https://github.com/rust-lang/rust/pull/79965 which is still on my todo list to clean up and resubmit with Windows code alongside the Unix code.

**Comment by the8472** [comment]
Do you have a motivating example where ENOSYS and EOPNOTSUPP can both occur on the same syscall and would require different handling?

**Comment by ijackson** [comment]
If you call `connect` and get `ENOSYS` it means that you are using the wrong syscall ABI, or the wrong binary has been installed for the platform.  Maybe the right answer is to exec a binary for a different kernel ABI.  If you get `EOPNOTSUPP` it means that the fd you are calling it on is not in the state you expected.  That might be an expected condition (eg, the socket is already connected).

**Comment by ijackson** [comment]
If you call `fchmodat` with `AT_SYMLINK_NOFOLLOW` and get `ENOSYS` you must fall back to some dance with `chdir` or something.  If you get `EOPNOTSUPP` it means the target was a symlink when you expected it not to be (probably for security reasons)  and the OS doesn't support symlink modesand the dance (if you went through it) would fail too.

**Comment by m-ou-se** [comment]
@rfcbot concern EOPNOTSUPP

See discusson starting at https://github.com/rust-lang/rust/pull/78880#issuecomment-777623233

**Comment by ijackson** [comment]
I am no expert on the wasi environmen but I would guess that WASI `ERRNO_NOSUP` is `EOPNOTSUPP` with a funny hat on.  In particular, that you get it from unexpected sequences of socket operations.

wasi is an environment where compatibility with different external environments, and fallback strategies, is more important than for native binaries.  So we should definitely try to DTRT there.

The best spec I found for what errno values there are, was this https://github.com/WebAssembly/WASI/blob/main/phases/snapshot/docs.md which bears a striking resemblance to a Unix errno list.

So probably whatever we do with `EOPNOTSUPP` (and `ENOTSUP`) should be done for wasi `ERRNO_NOSUP` too.

**Comment by the8472** [comment]
I'm not convinced by the given examples.

> If you call `connect` and get `ENOSYS` it means that you are using the wrong syscall ABI, or the wrong binary has been installed for the platform. Maybe the right answer is to exec a binary for a different kernel ABI. If you get `EOPNOTSUPP` it means that the fd you are calling it on is not in the state you the call expected. That might be an expected condition (eg, the socket is already connected).

Perhaps just ignorance on my end, but when would you expect to be able to run a program well enough to the point where you hit the wrong kernel ABI and then somehow recover from that problem? This doesn't seem like a recoverable thing you would deal with from within the program. You wouldn't even be able to print an error message or call exit/abort if that is misconfigured.

> If you call `fchmodat` with `AT_SYMLINK_NOFOLLOW` and get `ENOSYS` you must fall back to some dance with `chdir` or something. If you get `EOPNOTSUPP` it means the target was a symlink when you expected it not to be (probably for security reasons) and the OS doesn't support symlink modes and the dance (if you went through it) would fail too.

Does that apply to any supported rust platforms? At least on linux `fchmodat` was inttroduced in 2.6.16, which is older than the minimum supported version 2.6.32.

**Comment by ijackson** [comment]
> > If you call `connect` and get `ENOSYS` it means that you are using the wrong syscall ABI, or the wrong binary has been installed for the platform. Maybe the right answer is to exec a binary for a different kernel ABI. If you get `EOPNOTSUPP` it means that the fd you are calling it on is not in the state you the call expected. That might be an expected condition (eg, the socket is already connected).
> 
> Perhaps just ignorance on my end, but when would you expect to be able to run a program well enough to the point where you hit the wrong kernel ABI and then somehow recover from that problem?

There have in the past been kernel APIs where subsets of the system calls have been replaced with newer variants.  For example, large file support generally involved replacing calls to `fstat` et al.  You can see this as `lstat64` in strace on Linux.  Our API choices about errno handling are not just for now; they are for the future - both the future of majority platforms like Linux, and the future of Rust which we hope will become more widely ported.

In practice I agree that the most correct thing to do in most cases on getting `ENOSYS` is going to be to bail completely.  But even so in a program that is doing exciting socket stunts, that is a different appropriate response to the one for `EOPNOTSUPP`, which might well be to draw conclusions about the socket state and respond by making different socket calls.

Even in a very program, where the errors are handled by printing an error message and quitting, the error messages are different.  `ENOSYS` probably indicates a version incompatibility between the program and its environment, for which a message saying something about "unsupported" is appropriate.  The appropriate response of the system administrator is to double check the installation approach.

In a simple program, `EOPNOTSUPP` probably indicates a logical error in the program: it has made its socket calls in the wrong order, or the wrong combination, perhaps provoked by an unusual set of command line options or configuration settings.  The appropriate response for the user is probably to try to run the program a different way or to ask it to do something more simple.  The appropriate response for the programmer to the bug report is to look at the sequence of socket calls and fix the program's socket handling logic.

> > If you call `fchmodat` with `AT_SYMLINK_NOFOLLOW` and get `ENOSYS` you must fall back to some dance with `chdir` or something. If you get `EOPNOTSUPP` it means the target was a symlink when you expected it not to be (probably for security reasons) and the OS doesn't support symlink modes and the dance (if you went through it) would fail too.
> 
> Does that apply to any supported rust platforms? At least on linux `fchmodat` was inttroduced in 2.6.16, which is older than the minimum supported version 2.6.32.

I don't think we want to rule out supporting Rust on platforms without `fchmodat`!  So reasoning like this from the set of currently-supported platforms, to the proper treatment of Unix errnos forever more for all Rust programs, is not correct.

**Comment by the8472** [comment]
The examples haven't been particularly convincing but now I agree that there is a categorical distinction between those two regardless of current practical impact. The distinction is whether something is unconditionally unavailable (`ENOSYS`) or merely not applicable to a particular resource or in a particular state (`EOPNOTSUPP`). Although `EINVAL` is frequently used for the latter purpose too, e.g. when you pass the wrong kind of file descriptor to a syscall for example.

The doc comment already reflects this.

```rust
/// This operation is not supported on this platform.
```

**Comment by ijackson** [comment]
@the8472  Let us try to move forward; I'm not enjoying this debate very much and I guess neither are you :-/.

In practical terms for this MR, I think your last comment means you agree that `EOPNOTSUPP`, `ENOTSUP` and WASI's `ERRNO_NOSUP` ought to be dropped from the list of errno values for the new `ErrorKind::Unimplemented`?

**Comment by the8472** [comment]
> In practical terms for this MR, I think your last comment means you agree that `EOPNOTSUPP`, `ENOTSUP` and WASI's `ERRNO_NOSUP` ought to be dropped from the list of errno values for the new `ErrorKind::Unimplemented`?

Yes.

**Comment by CDirkx** [comment]
Looking deeper into the semantics of `EOPNOTSUPP` and `ENOTSUP`, I agree with the distinction from `ENOSYS`. The original motivation for including those error codes wasn't that strong anyway, and I am glad that people with more experience with UNIX errors than me weighed in.

To summarize the specific semantics in this proposal:
`std::io::ErrorKind::NotSupported` is an error returned when a given operation is not supported on a platform, and will thus never succeed; there is no way for the software to recover.

The only error codes that will map to `NotSupported`:
 - UNIX: `ENOSYS` "The return of [ENOSYS] is to be taken to indicate that the function of the interface is not supported at all; the function will always fail with this error code." ([POSIX.1-2001](http://www.open-std.org/jtc1/sc22/open/n4217.pdf#page=3452))
 - Windows: `ERROR_CALL_NOT_IMPLEMENTED` "This function is not supported on this system." ([Windows System Error Codes](https://docs.microsoft.com/en-us/windows/win32/debug/system-error-codes--0-499-))

That means I will drop the originally included error codes `EOPNOTSUPP` `ENOTSUP` `ERRNO_NOSUP` and `E_NOTIMPL`, as they all have meanings that do not directly correspond to `NotSupported`.

**Comment by m-ou-se** [comment]
@rfcbot resolve EOPNOTSUPP

**Comment by CDirkx** [comment]
I apparently missed some places `NotSupported` can be used, I added a commit to also include those.
I also updated `since` so this should now be ready to merge.

**Comment by joshtriplett** [comment]
I only just now saw this. I realize that it has already been through FCP, but I'm concerned about the name, specifically.

I don't think this is the right name for ENOSYS. It's already challenging, sometimes, to map these names to the underlying errno code. This name, in particular, seems much more likely to mean EOPNOTSUPP rather than ENOSYS, even though the semantic usage is like the latter, and Rust uses it more like the latter.

Sorry for the last-minute comment. I don't want to bikeshed naming, but I don't think this is a bikeshed concern. I think this is likely to produce confusion.

**Comment by CDirkx** [comment]
Maybe `Unsupported`? Like the `unsupported` function and `std::sys::unsupported`.

**Comment by joshtriplett** [comment]
@CDirkx Yeah, that seems fine. That's different enough that it doesn't *directly* sound like a direct mapping of a specific different error code. The rest could be handled via documentation.

**Comment by CDirkx** [comment]
I renamed `NotSupported` to `Unsupported` and added an extra bit of documentation: "This means that the operation can never succeed."

File: library/std/src/io/error.rs [hunk 3]
```diff
@@ -178,6 +178,12 @@ pub enum ErrorKind {
     /// read.
     #[stable(feature = "read_exact", since = "1.6.0")]
     UnexpectedEof,
+
+    /// This operation is unsupported on this platform.
+    ///
+    /// This means that the operation can never succeed.
+    #[stable(feature = "unsupported_error", since = "1.52.0")]
```
Comment (m-ou-se) [inline library/std/src/io/error.rs:185]: Nightly is at 1.53.0 now:
```suggestion
    #[stable(feature = "unsupported_error", since = "1.53.0")]
```

**Comment by CDirkx** [comment]
Merge conflicts have been resolved.

**Comment by CDirkx** [comment]
I fixed the clippy test that uses `ErrorKind`, but I'm not sure how the failure in `compile-fail/unaligned_pointers/reference_to_packed.rs` is related?

**Comment by jyn514** [comment]
@CDirkx hmm, try pushing your clippy fix and rebasing over master? I agree the compile-fail clippy test looks unrelated.

**Comment by CDirkx** [comment]
Rebased and fixed the clippy error, should be ready to try merging again.

**Comment by jyn514** [comment]
```
error: failed to download from `https://crates.io/api/v1/crates/minifier/0.0.39/download`
failed to run: /Users/runner/work/rust/rust/build/bootstrap/debug/bootstrap dist

Caused by:
  failed to get 200 response from `https://crates.io/api/v1/crates/minifier/0.0.39/download`, got 503
```
@bors retry

**Pull request merged**